### DISCUSSION TITLE
Issues with Running ROS Services[/cf1/takeoff] on command line

### DISCUSSION THREAD

**Discussion by tbaranoski** [USC-ACTLab/crazyswarm#440] [Q&A]
Thank you in advance,

I have the Crazyflie 2.1 and I am using Ubuntu 20.04 VM. I am using the flow deck with crazyswarm. 
When I run hover_swarm I am able to successfully takeoff using the start button on the Xbox controller as well as land using the back button. However, I am not bale to successfully use the ros services via the command line.

When I enter the command:
rosservice call /cf1/takeoff 0 .5 3

the crazyflie does nothing. However the command line shows:
[cf1] Takeoff

just like it does when I use the Xbox controller and hit the start button. Just not getting any movement at all by the Crazyflie.
I am wondering if I am missing something.

The log from the command line is shown below:

roslaunch crazyswarm hover_swarm.launch
... logging to /home/bitcraze/.ros/log/94fb56e8-ef0d-11eb-8a7d-25f561d364cb/roslaunch-ubuntu-7821.log
Checking log directory for disk usage. This may take a while.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://ubuntu:33629/

SUMMARY
========

PARAMETERS
 * /crazyflieTypes/CF21SingleMarker/batteryVoltageWarning: 3.8
 * /crazyflieTypes/CF21SingleMarker/batteryVoltateCritical: 3.7
 * /crazyflieTypes/CF21SingleMarker/bigQuad: False
 * /crazyflieTypes/CF21SingleMarker/dynamicsConfiguration: 0
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/i_range_m_z: 1500
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/i_range_xy: 2.0
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/i_range_z: 0.4
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/kR_xy: 70000
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/kR_z: 60000
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/kd_omega_rp: 200
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/kd_xy: 0.2
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/kd_z: 0.4
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/ki_m_z: 500
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/ki_xy: 0.05
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/ki_z: 0.05
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/kp_xy: 0.4
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/kp_z: 1.25
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/kw_xy: 20000
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/kw_z: 12000
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/mass: 0.032
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/ctrlMel/massThrust: 132000
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/kalman/mNGyro_rollpitch: 0.1
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/kalman/mNGyro_yaw: 0.1
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/kalman/pNAcc_xy: 1.0
 * /crazyflieTypes/CF21SingleMarker/firmwareParams/kalman/pNAcc_z: 2.0
 * /crazyflieTypes/CF21SingleMarker/markerConfiguration: 3
 * /crazyflieTypes/default/batteryVoltageWarning: 3.8
 * /crazyflieTypes/default/batteryVoltateCritical: 3.7
 * /crazyflieTypes/default/bigQuad: False
 * /crazyflieTypes/default/dynamicsConfiguration: 0
 * /crazyflieTypes/default/firmwareParams/ctrlMel/i_range_m_z: 1500
 * /crazyflieTypes/default/firmwareParams/ctrlMel/i_range_xy: 2.0
 * /crazyflieTypes/default/firmwareParams/ctrlMel/i_range_z: 0.4
 * /crazyflieTypes/default/firmwareParams/ctrlMel/kR_xy: 70000
 * /crazyflieTypes/default/firmwareParams/ctrlMel/kR_z: 60000
 * /crazyflieTypes/default/firmwareParams/ctrlMel/kd_omega_rp: 200
 * /crazyflieTypes/default/firmwareParams/ctrlMel/kd_xy: 0.2
 * /crazyflieTypes/default/firmwareParams/ctrlMel/kd_z: 0.4
 * /crazyflieTypes/default/firmwareParams/ctrlMel/ki_m_z: 500
 * /crazyflieTypes/default/firmwareParams/ctrlMel/ki_xy: 0.05
 * /crazyflieTypes/default/firmwareParams/ctrlMel/ki_z: 0.05
 * /crazyflieTypes/default/firmwareParams/ctrlMel/kp_xy: 0.4
 * /crazyflieTypes/default/firmwareParams/ctrlMel/kp_z: 1.25
 * /crazyflieTypes/default/firmwareParams/ctrlMel/kw_xy: 20000
 * /crazyflieTypes/default/firmwareParams/ctrlMel/kw_z: 12000
 * /crazyflieTypes/default/firmwareParams/ctrlMel/mass: 0.032
 * /crazyflieTypes/default/firmwareParams/ctrlMel/massThrust: 132000
 * /crazyflieTypes/default/markerConfiguration: 0
 * /crazyflieTypes/defaultSingleMarker/batteryVoltageWarning: 3.8
 * /crazyflieTypes/defaultSingleMarker/batteryVoltateCritical: 3.7
 * /crazyflieTypes/defaultSingleMarker/bigQuad: False
 * /crazyflieTypes/defaultSingleMarker/dynamicsConfiguration: 0
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/i_range_m_z: 1500
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/i_range_xy: 2.0
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/i_range_z: 0.4
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/kR_xy: 70000
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/kR_z: 60000
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/kd_omega_rp: 200
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/kd_xy: 0.2
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/kd_z: 0.4
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/ki_m_z: 500
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/ki_xy: 0.05
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/ki_z: 0.05
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/kp_xy: 0.4
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/kp_z: 1.25
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/kw_xy: 20000
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/kw_z: 12000
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/mass: 0.032
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/ctrlMel/massThrust: 132000
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/kalman/mNGyro_rollpitch: 0.2
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/kalman/mNGyro_yaw: 0.1
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/kalman/pNAcc_xy: 4.0
 * /crazyflieTypes/defaultSingleMarker/firmwareParams/kalman/pNAcc_z: 8.0
 * /crazyflieTypes/defaultSingleMarker/markerConfiguration: 3
 * /crazyflieTypes/large/batteryVoltageWarning: 11.4
 * /crazyflieTypes/large/batteryVoltateCritical: 11.1
 * /crazyflieTypes/large/bigQuad: True
 * /crazyflieTypes/large/dynamicsConfiguration: 0
 * /crazyflieTypes/large/firmwareParams/ctrlMel/i_range_m_z: 1500
 * /crazyflieTypes/large/firmwareParams/ctrlMel/i_range_xy: 2.0
 * /crazyflieTypes/large/firmwareParams/ctrlMel/i_range_z: 4.0
 * /crazyflieTypes/large/firmwareParams/ctrlMel/kR_xy: 30000
 * /crazyflieTypes/large/firmwareParams/ctrlMel/kR_z: 40000
 * /crazyflieTypes/large/firmwareParams/ctrlMel/kd_omega_rp: 100
 * /crazyflieTypes/large/firmwareParams/ctrlMel/kd_xy: 3.0
 * /crazyflieTypes/large/firmwareParams/ctrlMel/kd_z: 3.0
 * /crazyflieTypes/large/firmwareParams/ctrlMel/ki_m_z: 500
 * /crazyflieTypes/large/firmwareParams/ctrlMel/ki_xy: 1.0
 * /crazyflieTypes/large/firmwareParams/ctrlMel/ki_z: 2.0
 * /crazyflieTypes/large/firmwareParams/ctrlMel/kp_xy: 8.0
 * /crazyflieTypes/large/firmwareParams/ctrlMel/kp_z: 8.0
 * /crazyflieTypes/large/firmwareParams/ctrlMel/kw_xy: 9000
 * /crazyflieTypes/large/firmwareParams/ctrlMel/kw_z: 10000
 * /crazyflieTypes/large/firmwareParams/ctrlMel/mass: 0.38
 * /crazyflieTypes/large/firmwareParams/ctrlMel/massThrust: 5400
 * /crazyflieTypes/large/markerConfiguration: 2
 * /crazyflieTypes/medium/batteryVoltageWarning: 7.6
 * /crazyflieTypes/medium/batteryVoltateCritical: 7.4
 * /crazyflieTypes/medium/bigQuad: True
 * /crazyflieTypes/medium/dynamicsConfiguration: 0
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/i_range_m_z: 1500
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/i_range_xy: 2.0
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/i_range_z: 2.0
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/kR_xy: 50000
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/kR_z: 60000
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/kd_omega_rp: 200
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/kd_xy: 1.0
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/kd_z: 1.0
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/ki_m_z: 500
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/ki_xy: 0.1
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/ki_z: 0.5
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/kp_xy: 2.0
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/kp_z: 3
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/kw_xy: 16000
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/kw_z: 12000
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/mass: 0.11
 * /crazyflieTypes/medium/firmwareParams/ctrlMel/massThrust: 23000
 * /crazyflieTypes/medium/markerConfiguration: 1
 * /crazyflies: [{'channel': 100,...
 * /crazyswarm_server/enable_logging: True
 * /crazyswarm_server/enable_parameters: True
 * /crazyswarm_server/firmwareParams/commander/enHighLevel: 1
 * /crazyswarm_server/firmwareParams/kalman/resetEstimation: 1
 * /crazyswarm_server/firmwareParams/locSrv/extPosStdDev: 1e-3
 * /crazyswarm_server/firmwareParams/locSrv/extQuatStdDev: 0.05
 * /crazyswarm_server/firmwareParams/ring/effect: 16
 * /crazyswarm_server/firmwareParams/ring/headlightEnable: 0
 * /crazyswarm_server/firmwareParams/ring/solidBlue: 255
 * /crazyswarm_server/firmwareParams/ring/solidGreen: 0
 * /crazyswarm_server/firmwareParams/ring/solidRed: 0
 * /crazyswarm_server/firmwareParams/stabilizer/controller: 2
 * /crazyswarm_server/firmwareParams/stabilizer/estimator: 2
 * /crazyswarm_server/force_no_cache: False
 * /crazyswarm_server/genericLogTopicFrequencies: [10]
 * /crazyswarm_server/genericLogTopic_log1_Variables: ['stateEstimate.x...
 * /crazyswarm_server/genericLogTopics: ['log1']
 * /crazyswarm_server/motion_capture_type: none
 * /crazyswarm_server/object_tracking_type: libobjecttracker
 * /crazyswarm_server/print_latency: False
 * /crazyswarm_server/save_point_clouds: ~/pointCloud.ot
 * /crazyswarm_server/send_position_only: False
 * /crazyswarm_server/vicon_host_name: vicon
 * /crazyswarm_server/world_frame: /world
 * /crazyswarm_server/write_csvs: False
 * /crazyswarm_teleop/csv_file: /home/bitcraze/cr...
 * /crazyswarm_teleop/timescale: 0.8
 * /dynamicsConfigurations/0/maxFitnessScore: 0.001
 * /dynamicsConfigurations/0/maxPitch: 1.4
 * /dynamicsConfigurations/0/maxPitchRate: 20.0
 * /dynamicsConfigurations/0/maxRoll: 1.4
 * /dynamicsConfigurations/0/maxRollRate: 20.0
 * /dynamicsConfigurations/0/maxXVelocity: 2.0
 * /dynamicsConfigurations/0/maxYVelocity: 2.0
 * /dynamicsConfigurations/0/maxYawRate: 10.0
 * /dynamicsConfigurations/0/maxZVelocity: 3.0
 * /joy/dev: /dev/input/js2
 * /markerConfigurations/0/numPoints: 4
 * /markerConfigurations/0/offset: [0.0, -0.01, -0.04]
 * /markerConfigurations/0/points/0: [0.0177184, 0.013...
 * /markerConfigurations/0/points/1: [-0.0262914, 0.05...
 * /markerConfigurations/0/points/2: [-0.0328889, -0.0...
 * /markerConfigurations/0/points/3: [0.0431307, -0.03...
 * /markerConfigurations/1/numPoints: 4
 * /markerConfigurations/1/offset: [0.0, 0.0, -0.03]
 * /markerConfigurations/1/points/0: [-0.00896228, -0....
 * /markerConfigurations/1/points/1: [-0.0156318, 0.09...
 * /markerConfigurations/1/points/2: [0.0461693, -0.08...
 * /markerConfigurations/1/points/3: [-0.0789959, -0.0...
 * /markerConfigurations/2/numPoints: 4
 * /markerConfigurations/2/offset: [0.0, 0.0, -0.06]
 * /markerConfigurations/2/points/0: [0.0558163, -0.00...
 * /markerConfigurations/2/points/1: [-0.0113941, 0.00...
 * /markerConfigurations/2/points/2: [-0.0306277, 0.05...
 * /markerConfigurations/2/points/3: [0.0535816, -0.04...
 * /markerConfigurations/3/numPoints: 1
 * /markerConfigurations/3/offset: [0.0, -0.01, -0.04]
 * /markerConfigurations/3/points/0: [0.0177184, 0.013...
 * /numDynamicsConfigurations: 1
 * /numMarkerConfigurations: 4
 * /rosdistro: noetic
 * /rosversion: 1.15.11

NODES
  /
    crazyswarm_server (crazyswarm/crazyswarm_server)
    crazyswarm_teleop (crazyswarm/crazyswarm_teleop)
    joy (joy/joy_node)
    rviz (rviz/rviz)

auto-starting new master
process[master]: started with pid [7836]
ROS_MASTER_URI=http://localhost:11311

setting /run_id to 94fb56e8-ef0d-11eb-8a7d-25f561d364cb
process[rosout-1]: started with pid [7851]
started core service [/rosout]
process[crazyswarm_server-2]: started with pid [7858]
process[joy-3]: started with pid [7859]
process[crazyswarm_teleop-4]: started with pid [7860]
process[rviz-5]: started with pid [7861]
[ INFO] [1627412719.408148218]: Wait for services...
[ INFO] [1627412719.409320783]: waitForService: Service [/emergency] has not been advertised, waiting...
[ WARN] [1627412719.419132161]: Couldn't set gain on joystick force feedback: Bad file descriptor
[ INFO] [1627412719.420516095]: Opened joystick: /dev/input/js2 (Microsoft X-Box One pad). deadzone_: 0.050000.
[ INFO] [1627412719.430164868]: waitForService: Service [/emergency] is now available.
[ INFO] [1627412719.431456666]: Manager ready.
ch: 1
[ INFO] [1627412719.525961971]: Adding CF: cf1 (radio://0/100/2M/E7E7E7E701, cf1)...
[ INFO] [1627412719.586673562]: CF ctor: 0.060008 s
[ INFO] [1627412719.619292353]: [cf1] SYS: ----------------------------
[ INFO] [1627412719.642727362]: [cf1] SYS: Crazyflie 2.1 is up and running!
[ INFO] [1627412719.665350493]: [cf1] SYS: Build 11:0864ef92245a (2021.03 +11) CLEAN
[ INFO] [1627412719.700941520]: [cf1] SYS: I am 0x20383347344D50080030004D and I have 1024KB of flash!
[ INFO] [1627412719.711080614]: [cf1] CFGBLK: v1, verification [OK]
[ INFO] [1627412719.725537028]: [cf1] DECK_CORE: 2 deck(s) found
[ INFO] [1627412719.747111883]: [cf1] DECK_CORE: Calling INIT on driver bcFlow2 for deck 0
[ INFO] [1627412719.758969473]: [cf1] ZR2: Z-down sensor [OK]
[ INFO] [1627412719.780328371]: [cf1] PMW: Motion chip id: 0x49:0xB6
[ INFO] [1627412719.800788133]: [cf1] DECK_CORE: Calling INIT on driver bcMultiranger for deck 1
[ INFO] [1627412719.824223347]: [cf1] IMU: BMI088: Using I2C interface.
[ INFO] [1627412719.846678041]: [cf1] IMU: BMI088 Gyro connection [OK].
[ INFO] [1627412719.868173797]: [cf1] IMU: BMI088 Accel connection [OK]
[ INFO] [1627412719.891523323]: [cf1] IMU: BMP388 I2C connection [OK]
[ INFO] [1627412719.912070839]: [cf1] ESTIMATOR: Using Kalman (2) estimator
[ INFO] [1627412719.934496389]: [cf1] CONTROLLER: Using PID (1) controller
[ INFO] [1627412719.955997220]: [cf1] MTR-DRV: Using brushed motor driver
[ INFO] [1627412719.966662847]: [cf1] EEPROM: I2C connection [OK].
[ INFO] [1627412719.980075114]: [cf1] STORAGE: Storage check [OK].
[ INFO] [1627412720.003235721]: [cf1] IMU: BMI088 gyro self-test [OK]
[ INFO] [1627412720.025421255]: [cf1] MR: Init front sensor [OK]
[ INFO] [1627412720.035983991]: [cf1] MR: Init back sensor [OK]
[ INFO] [1627412720.044877372]: [cf1] MR: Init up sensor [OK]
[ INFO] [1627412720.056478111]: [cf1] MR: Init left sensor [OK]
[ INFO] [1627412720.067241679]: [cf1] MR: Init right sensor [OK]
[ INFO] [1627412720.079011359]: [cf1] DECK_CORE: Deck 1 test [OK].
[ INFO] [1627412720.099468347]: [cf1] ion...
[ INFO] [1627412720.111175007]: [cf1] SYS: Free heap: 16352 bytes
[ INFO] [1627412720.123850852]: [cf1] STAB: Ready to fly.
[ INFO] [1627412720.142485970]: [cf1] ESTKALMAN: State out of bounds, resetting
[ INFO] [1627412720.168779151]: [cf1] Requesting parameters...
[ INFO] [1627412720.581961298]: Found variables in cache.
[ INFO] [1627412723.304284155]: [cf1] reqParamTOC: 3.717500 s (229 params)
[ INFO] [1627412723.304308593]: [cf1] Requesting logging variables...
[ INFO] [1627412723.335197734]: Found variables in cache.
[ INFO] [1627412723.335792025]: [cf1] reqLogTOC: 0.031513 s
[ INFO] [1627412723.401672969]: [cf1] logBlocks: 3.717500 s
[ INFO] [1627412723.476561380]: Requesting memories...
[ INFO] [1627412723.573817616]: Memories: 6
[ INFO] [1627412723.867998171]: [cf1] Ready. Elapsed: 4.281203 s
[ INFO] [1627412723.868034145]: CF run: 4.281370 s
update commander/enHighLevel to 1
update kalman/resetEstimation to 1
update locSrv/extPosStdDev to 0.001
update locSrv/extQuatStdDev to 0.05
update ring/effect to 16
update ring/headlightEnable to 0
update ring/solidBlue to 255
update ring/solidGreen to 0
update ring/solidRed to 0
update stabilizer/controller to 2
update stabilizer/estimator to 2
update ctrlMel/i_range_m_z to 1500
update ctrlMel/i_range_xy to 2
update ctrlMel/i_range_z to 2
update ctrlMel/kR_xy to 50000
update ctrlMel/kR_z to 60000
update ctrlMel/kd_omega_rp to 200
update ctrlMel/kd_xy to 1
update ctrlMel/kd_z to 1
update ctrlMel/ki_m_z to 500
update ctrlMel/ki_xy to 0.1
update ctrlMel/ki_z to 0.5
update ctrlMel/kp_xy to 2
update ctrlMel/kp_z to 3
update ctrlMel/kw_xy to 16000
update ctrlMel/kw_z to 12000
update ctrlMel/mass to 0.11
update ctrlMel/massThrust to 23000
[ INFO] [1627412723.880822417]: [cf1] Update parameters
[ INFO] [1627412724.527607017]: [cf1] CONTROLLER: Using Mellinger (2) controller
[ INFO] [1627412724.943110810]: Update params: 1.075039 s
[ INFO] [1627412724.943432905]: Started 1 threads
[ INFO] [1627412734.115047933]: [cf1] Takeoff

**Comment by jpreiss** [comment]
I never used `rosservice call` with Crazyswarm, but I don't see any reason why it wouldn't work.

I can only suggest going the normal software debugging route. Since you get the `[cf1] Takeoff` log message, I would add some print statements in `CrazyflieROS::takeoff` to validate the argument parsing, and use a debugger if needed.

**Comment by whoenig** [comment]
I believe the problem is the third parameter (duration): this is not a simple number, but a composite type (secs/nsecs), see http://wiki.ros.org/msg. Thus, you'll need to construct it using the YAML syntax instead, see https://wiki.ros.org/ROS/YAMLCommandLine.

**Reply by tbaranoski** [reply]
I got it working, thank you!
The documentation doesn't exactly spell it out, so in case this helps someone else the command I entered was:

rosservice call /cf1/takeoff '{groupMask: 0, height: .5, duration: { secs: 3, nsecs: 5}'}
